### PR TITLE
Stunbatons hit only the first person when thrown

### LINKED ISSUE #19124: Stamina damage on hit should only be on first mob
Hitting every mob in line is bad.

### PR DESCRIPTION
**Pull request by kalane15**
## About the PR

Fixed stunnig all mobs in a line when thrown
Closes #19124.
**Media**

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- tweak: Stunbatons now hit only the first person when thrown.

### CODE REVIEW

**Comment by deltanedas** [comment]
why close bowling is awful

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Better to just make the entity land after ThrowDoHitEvent is handled instead of this.

**Comment by kalane15** [comment]
This should affect all throwable items or stunbatons only? I'm not sure about it.

**Comment by metalgearsloth** [comment]
> Should this affect all throwable items or stunbatons only? I'm not sure about it.

All of them.

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 1]
```diff
@@ -219,17 +222,27 @@ private void OnProjectileHit(EntityUid uid, StaminaDamageOnCollideComponent comp
...
-        OnCollide(uid, component, args.Target);
+        if (TryComp<ThrownItemComponent>(uid, out var thrownComp) &&
+            HasComp<StaminaComponent>(args.Target))
+        {
+            if (OnCollide(uid, component, args.Target))
+            {
+                _thrownItemSystem.StopThrow(uid, thrownComp);
+            }
+        }
     }
```
Comment (metalgearsloth) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:233]: Add a field to ThrowDoHitEvent on whether the entity should land.

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 2]
```diff
@@ -219,17 +222,27 @@ private void OnProjectileHit(EntityUid uid, StaminaDamageOnCollideComponent comp
...
+            return false;
+
         var ev = new StaminaDamageOnHitAttemptEvent();
         RaiseLocalEvent(uid, ref ev);
         if (ev.Cancelled)
-            return;
+            return false;
 
-        TakeStaminaDamage(target, component.Damage, source: uid, sound: component.Sound);
+        return TakeStaminaDamage(target, component.Damage, staminaComp, uid, sound: component.Sound);
```
Comment (metalgearsloth) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:245]: You're resolving staminacomponent twice which is unnecessary, just pass it into the method as it's far cheaper.

Also call it TryCollide instead if you're changing the return type.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Throwing/ThrownItemSystem.cs [hunk 3]
```diff
@@ -144,7 +144,7 @@ public void ThrowCollideInteraction(ThrownItemComponent component, EntityUid thr
                     $"{ToPrettyString(thrown):thrown} thrown by {ToPrettyString(component.Thrower.Value):thrower} hit {ToPrettyString(target):target}.");
 
             RaiseLocalEvent(target, new ThrowHitByEvent(thrown, target, component), true);
-            RaiseLocalEvent(thrown, new ThrowDoHitEvent(thrown, target, component), true);
+            RaiseLocalEvent(thrown, new ThrowDoHitEvent(thrown, target, component, true), true);
```
Comment (metalgearsloth) [inline Content.Shared/Throwing/ThrownItemSystem.cs:147]: LandAfterHit is an unnecessary field.

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 4]
```diff
@@ -268,15 +301,6 @@ public bool TryTakeStamina(EntityUid uid, float value, StaminaComponent? compone
...
 
-        var ev = new BeforeStaminaDamageEvent(value);
-        RaiseLocalEvent(uid, ref ev);
-        if (ev.Cancelled)
-            return;
-
-        // Have we already reached the point of max stamina damage?
-        if (component.Critical)
-            return;
-
```
Comment (metalgearsloth) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:279]: If someone's crit they shouldn't be taking more stamina damage so this should stay.

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 5]
```diff
@@ -244,21 +257,41 @@ private void SetStaminaAlert(EntityUid uid, StaminaComponent? component = null)
...
 
         var oldStam = component.StaminaDamage;
 
+        // Have we already reached the point of max stamina damage?
         if (oldStam + value > component.CritThreshold || component.Critical)
             return false;
 
-        TakeStaminaDamage(uid, value, component, source, with, visual: false);
+       TakeStaminaDamage(uid, value, component, source, with, visual: false, sound);
         return true;
```
Comment (metalgearsloth) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:295]: Caller should be handling sound, not stamina system.

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 6]
```diff
@@ -244,21 +257,41 @@ private void SetStaminaAlert(EntityUid uid, StaminaComponent? component = null)
...
+            return false;
+
+        var ev = new BeforeStaminaDamageEvent(value);
+        RaiseLocalEvent(uid, ref ev);
+        if (ev.Cancelled)
+            return false;
+
+        TakeStaminaDamage(uid, value, component, source, with, visual: false, sound);
+        return true;
+    }
```
Comment (metalgearsloth) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:275]: Stamina shouldn't be tied to mob component.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
@kalane15 are you still working on this

**Comment by kalane15** [comment]
unfortunately no

**Pull request closed without merging**